### DISCUSSION TITLE
How long should it take for both sides of a link to recognize disconnects?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [markqvist/Reticulum#319] [Help & Questions] [answered]
I have been running tests with both sides connected via a link over TCP, and it takes a minimum of 20 minutes (sometimes 30 - 40) for either side to fire off the link closed callback when RNS.Reticulum.exit_handler() is called.

**Comment by markqvist** [comment]
Are you calling `RNS.Reticulum.exit_handler()` yourself? You shouldn't do that.

**Reply by ThomasFreedman** [reply]
Ok, so how then should I be shutting down reticulum cleanly, by simply using del on the class that creates the instance? And with such a direct connecction, are such delays normal?

**Reply by markqvist** [reply, marked as answer]
> Ok, so how then should I be shutting down reticulum cleanly

Just do:

```python
exit(0) # Or whatever non-0 exit code you need
```

That's all. Reticulum detects the program termination and cleans up. Any open links should be closed cleanly and automatically send teardown packets.

The problem you are seeing is more or less completely equivalent to immediately terminating a program with an open TCP connection. It will take a long time for the remote host to detect that the connection is no longer active (unless certain TCP parameters were set on the socket via the kernel interfaces).

**Reply by ThomasFreedman** [reply]
Ok, so what I'm seeing is normal then. Thanks for the clarification. 

I'm writing a new question about resource use I'll detail momentarily. In a nutshell, I don't need resources to stream, but rather just overcome the packet size restriction. Just not sure about the callbacks & payload delivery. Seems like I need to handle packet payloads differently than resource payloads, but not sure. Trying to follow the file transfer example but that's more about streaming file data.
